stackslib/src/net/api/getsortition.rs

```diff
@@ -0,0 +1,380 @@
+    /// the tenure that this new burn block's miner will be building off of.
+    #[serde(with = "prefix_opt_hex")]
+    pub stacks_parent_ch: Option<ConsensusHash>,
+    /// If sortition occurred, this will be the consensus hash of the most recent sortition before
+    ///  this one.
+    #[serde(with = "prefix_opt_hex")]
+    pub last_sortition_ch: Option<ConsensusHash>,
+    #[serde(with = "prefix_opt_hex")]
+    /// In Stacks 2.x, this is the winning block.
+    /// In Stacks 3.x, this is the first block of the parent tenure.
+    pub committed_block_hash: Option<BlockHeaderHash>,
+}
+
+mod prefix_opt_hex {
+    pub fn serialize<S: serde::Serializer, T: std::fmt::Display>(
+        val: &Option<T>,
+        s: S,
+    ) -> Result<S::Ok, S::Error> {
+        match val {
+            Some(ref some_val) => {
+                let val_str = format!("0x{some_val}");
+                s.serialize_some(&val_str)
+            }
+            None => s.serialize_none(),
+        }
+    }
+
+    pub fn deserialize<'de, D: serde::Deserializer<'de>, T: super::HexDeser>(
+        d: D,
+    ) -> Result<Option<T>, D::Error> {
+        let opt_inst_str: Option<String> = serde::Deserialize::deserialize(d)?;
+        let Some(inst_str) = opt_inst_str else {
+            return Ok(None);
+        };
+        let Some(hex_str) = inst_str.get(2..) else {
+            return Err(serde::de::Error::invalid_length(
+                inst_str.len(),
+                &"at least length 2 string",
+            ));
+        };
+        let val = T::try_from(&hex_str).map_err(serde::de::Error::custom)?;
+        Ok(Some(val))
+    }
+}
+
+mod prefix_hex {
+    pub fn serialize<S: serde::Serializer, T: std::fmt::Display>(
+        val: &T,
+        s: S,
+    ) -> Result<S::Ok, S::Error> {
+        s.serialize_str(&format!("0x{val}"))
+    }
+
+    pub fn deserialize<'de, D: serde::Deserializer<'de>, T: super::HexDeser>(
+        d: D,
+    ) -> Result<T, D::Error> {
+        let inst_str: String = serde::Deserialize::deserialize(d)?;
+        let Some(hex_str) = inst_str.get(2..) else {
+            return Err(serde::de::Error::invalid_length(
+                inst_str.len(),
+                &"at least length 2 string",
+            ));
+        };
+        T::try_from(&hex_str).map_err(serde::de::Error::custom)
+    }
+}
+
+trait HexDeser: Sized {
+    fn try_from(hex: &str) -> Result<Self, HexError>;
+}
+
+macro_rules! impl_hex_deser {
+    ($thing:ident) => {
+        impl HexDeser for $thing {
+            fn try_from(hex: &str) -> Result<Self, HexError> {
+                $thing::from_hex(hex)
+            }
+        }
+    };
+}
+
+impl_hex_deser!(BurnchainHeaderHash);
+impl_hex_deser!(SortitionId);
+impl_hex_deser!(ConsensusHash);
+impl_hex_deser!(BlockHeaderHash);
+impl_hex_deser!(Hash160);
+
+impl TryFrom<(&String, &String)> for QuerySpecifier {
+    type Error = Error;
+
+    fn try_from(value: (&String, &String)) -> Result<Self, Self::Error> {
```

Comment:
I'm not really a fan of API endpoints that take query parameters for a small-domain set of inputs, especially when it allows for nonsensical queries like `?consensus=0x123&burn_height=4`.

Can this endpoint instead be structured as `/v3/sortitions/{identifier_name}/{identifier_value}`?  You'd support the following:

* `/v3/sortitions/consensus/{consensus_hash}`
* `/v3/sortitions/burn/{burn_hash}`
* `/v3/sortitions/burn_height/{burn_height}`

Then, there's exactly one correct way to issue a sortition query.

Reply:
This endpoint already only allows exactly one query parameter to be provided. The only difference in what you describe and this is whether or not we rely on a regex parser to perform this extended matching. I can update it to remove the query parameter though because I’m pretty ambivalent between those options.

Reply:
This endpoint allows many query parameters; it merely chooses the last one as the parameter it acts upon.  I personally think this is surprising -- I'd expect a nonsensical query to result in an HTTP error.  But if we want the property that this acts on at most one query, then coding that directly into the request path I think is the more "RESTful" way of going about it.

Reply:
Updated in fdcaea8b91972ad721d9d9e0579ac89fb531ddc0 to use the request path.

(FWIW, the query parameter _did_ enforce the caller supplying exactly 1 or 0 parameters -- it yielded a 4xx error otherwise)